Port DarkKronicle's bedwars display

moehreag opened the pull request:
Original PR: https://github.com/Sol-Client/client/pull/136

### Other changes
- Add more Events

### TODO
- [x] test the functionality, make changes as desired
- [x] Add the translation keys
- [x] port to all other versions
- [x] port the Event "system" to all other versions

moehreag committed:
port https://github.com/Sol-Client/client/pull/136

moehreag committed:
Merge branch 'multiversion' into feat/sol/darkkronicle-bedwars-display

moehreag committed:
Merge branch 'multiversion' into feat/sol/darkkronicle-bedwars-display

moehreag committed:
remove unnecessary old code

moehreag committed:
move class

moehreag committed:
restore functionality

moehreag committed:
apply license header

moehreag committed:
fix some bugs, add translations

This should now make the PR be in the same state as the original PR

moehreag committed:
fully migrate to new Event system

moehreag committed:
port to the other versions

some mixins might have issues, needs testing

moehreag committed:
remove timings code, fix some mixins

moehreag committed:
Merge branch 'multiversion' into feat/sol/darkkronicle-bedwars-display

moehreag committed:
oversaw this

moehreag committed:
fix more bugs

moehreag committed:
add more customisation options

- move common stuff to the common module

moehreag committed:
format a lot of things

- put the top 3 killers in the bottom of the player list

moehreag committed:
add customisable death messages

moehreag committed:
add credits

moehreag commented:
Versions other than 1.8.9 are still mostly untested.

TheKodeToad commented:
Sybillian didn't give permission for this project which doesn't feel great.

moehreag commented:
true, we can currently only rely on what's given on the [original github page](https://github.com/Sybillian/bw-overlay)

> This pack was created by Sybillian. Please clearly credit if you use it as a base for other projects.

The author still seems to be active in the community and on the hypixel forums. It might be worth contacting them for certainty.

moehreag committed:
fix a bug, add option to usually hide the hud

moehreag committed:
re-use original footer

- modify credits for Sybillian to match what is probably expected by them

moehreag committed:
use vanilla textures instead of custom ones

- advantages: easy replacement through resourcepacks
- disadvantages: *slightly* more messy code

moehreag committed:
update credits to reflect the latest changes

moehreag commented:
> Sybillian didn't give permission for this project which doesn't feel great.

this has now also been solved by just not needing any external textures anymore, but it might objectively look slightly worse now.

moehreag committed:
oversight

moehreag committed:
fix a potential crash

moehreag committed:
remove unnecessary darkening of the disabled furnace

- this isn't shown anyway but still

moehreag committed:
fix typo + PlayerHud scaling

moehreag committed:
add option to remove vignette, fix more things

- f.e. interpreting sun.properties files as skyX.properties

moehreag committed:
apply sky fix on other versions as well

moehreag committed:
another spotted error + missing translation key

moehreag committed:
suggested changes by DarkKronicle

moehreag committed:
mixin fixes + remove unused methods

moehreag committed:
change bottom bar

TheKodeToad commented:
Okay... idk how I'm going to review all this :sob:
14k lines is a lot :P

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/HypixelAbstractionLayer.java:
```diff
@@ -85,6 +94,19 @@ public static int getPlayerLevel(String uuid, String mode) {
 		return 0;
 	}
 
+	private static PlayerReply.Player getPlayer(String uuid) {
```
I think this should be used in other places :3

moehreag replied:
Which places do you imagine this could be used in? And remember that because the Hypixel API library is relocated, it cannot be referenced in the versioned modules.

TheKodeToad replied:
in the same class

moehreag replied:
done!

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/HypixelAbstractionLayer.java:
```diff
@@ -62,6 +64,13 @@ public static boolean hasValidAPIKey() {
+		if (loadPlayerDataIfAbsent(uuid)) {
+			return getPlayer(uuid).getProperty(stat);
+		}
+		return new JsonObject();
```
Considering the single place you use this has a null check, why not return null or an empty optional?

moehreag replied:
done!

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsLevelHeadMode.java:
```diff
@@ -0,0 +1,67 @@
+
+	static {
+		modes = new HashMap<>();
+		Arrays.stream(values()).forEach(m -> modes.put(m.toString(), m));
```
```suggestion
		Arrays.stream(values()).forEach(m -> modes.put(m.name(), m));
```

slightly safer

moehreag replied:
they both have the same method body, but also done.

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsLevelHeadMode.java:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	public static BedwarsLevelHeadMode get(String mode) {
+		return modes.getOrDefault(mode, null);
```
```suggestion
		return modes.get(mode);
```

I think this works too, unless I forgot something

moehreag replied:
yes, it does the exact same thing.

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsLevelHeadMode.java:
```diff
@@ -0,0 +1,67 @@
+		return titleSupplier.apply(stats);
+	}
+
+	private static final HashMap<String, BedwarsLevelHeadMode> modes;
```
```suggestion
	private static final Map<String, BedwarsLevelHeadMode> modes;
```
usually preferred

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsMessages.java:
```diff
@@ -0,0 +1,312 @@
+		"{killed} was chased away by {player}'s Golem."
+	);
+
+	public final static Pattern[] BED_BREAK = {
+		Pattern.compile(formatPlaceholder("Bed was broken by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was incinerated by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was iced by {player}")),
+		Pattern.compile(formatPlaceholder("Bed had to raise the white flag to {player}")),
+		Pattern.compile(formatPlaceholder("Bed was dismantled by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was deep fried by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was ripped apart by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was traded in for milk and cookies by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was sacrificed by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was gulped by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was gulped by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was squeaked apart by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was stung by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was impaled by {player}")),
+		Pattern.compile(formatPlaceholder("Bed be shot with cannon by {player}")),
+		Pattern.compile(formatPlaceholder("Bed got memed by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was made into a snowman by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was scrambled by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was stuffed with tissue paper by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was scrambled by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was bed #{number} destroyed by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was spooked by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was dreadfully corrupted by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was bed #{number} destroyed by {player}")),
+		Pattern.compile(formatPlaceholder("Bed exploded from a firework by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was blasted to dust by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was melted by {player}'s holiday spirit")),
+		Pattern.compile(formatPlaceholder("Bed was ripped to shreds by {player}")),
+		Pattern.compile(formatPlaceholder("Bed has left the game after seeing {player}")),
+		Pattern.compile(formatPlaceholder("Bed was spooked by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was contaminated by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was sold in a garage sale by {player}")),
+		Pattern.compile(formatPlaceholder("Bed was destroyed by {player}")),
+	};
```
```suggestion
	public final static Pattern[] BED_BREAK = Arrays.stream(
		"Bed was broken by {player}",
		"Bed was incinerated by {player}",
		"Bed was iced by {player}",
		"Bed had to raise the white flag to {player}",
		"Bed was dismantled by {player}",
		"Bed was deep fried by {player}",
		"Bed was ripped apart by {player}",
		"Bed was traded in for milk and cookies by {player}",
		"Bed was sacrificed by {player}",
		"Bed was gulped by {player}",
		"Bed was gulped by {player}",
		"Bed was squeaked apart by {player}",
		"Bed was stung by {player}",
		"Bed was impaled by {player}",
		"Bed be shot with cannon by {player}",
		"Bed got memed by {player}",
		"Bed was made into a snowman by {player}",
		"Bed was scrambled by {player}",
		"Bed was stuffed with tissue paper by {player}",
		"Bed was scrambled by {player}",
		"Bed was bed #{number} destroyed by {player}",
		"Bed was spooked by {player}",
		"Bed was dreadfully corrupted by {player}",
		"Bed was bed #{number} destroyed by {player}",
		"Bed exploded from a firework by {player}",
		"Bed was blasted to dust by {player}",
		"Bed was melted by {player}'s holiday spirit",
		"Bed was ripped to shreds by {player}",
		"Bed has left the game after seeing {player}",
		"Bed was spooked by {player}",
		"Bed was contaminated by {player}",
		"Bed was sold in a garage sale by {player}",
		"Bed was destroyed by {player}"
	).map(BedwarsMessages::formatPlaceholder).map(Pattern::compile).toArray(Pattern[]::new);
```

moehreag replied:
done!

TheKodeToad commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsPlayerStats.java:
```diff
@@ -0,0 +1,170 @@
+
+
+	public static BedwarsPlayerStats generateFake() {
+		return new BedwarsPlayerStats(
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			0,
+			1
+		);
```
```suggestion
		return new BedwarsPlayerStats(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1);
```

moehreag replied:
done!

DarkKronicle replied:
To be compliant with hypixel API terms this needs to generate fake stats so it can't be used to test if someone has a nick. I can come up with the algorithm

moehreag replied:
okay!

TheKodeToad commented on gradle.properties:
```diff
@@ -2,12 +2,10 @@
-
 # Mod Properties
-version=3.0.3
+version=3.0.4-darkkronicle-bedwars-overlay-beta.7
```
I don't think you want this!

TheKodeToad commented on 1.20/src/main/java/io/github/axolotlclient/config/screen/CreditsScreen.java:
```diff
@@ -161,11 +161,11 @@ private void initCredits() {
 		credits.add(new Credit("YakisikliBaran", "Turkish Translation"));
 		credits.add(new Credit("TheKodeToad", "Contributor", "Motion Blur", "Freelook", "Zoom"));
 		credits.add(new Credit("DragonEggBedrockBreaking", "Bugfixing", "Inspiration of new Features"));
+		credits.add(new Credit("DarkKronicle", "Bedwars Overlay", "Author of KronHUD, the best HUD mod!"));
```
I think darkkronicle should be above dragoneggbedrockbreaking as while they haven't directly contributed, the code you have used from him is pretty major :laughing:

moehreag replied:
I don't think that this list was sorted in any way, rather just appended when someone new contributed. If this is important to you or anyone else, it can be changed.

TheKodeToad replied:
You moved DarkKronicle

moehreag replied:
from the "Other people" list to the "Contributors" list (i think)

TheKodeToad replied:
ah... i wonder if this makes sense... but i understand

moehreag replied:
It might be clearer on the API rewrite branch where the Credits data has been put into its own class

TheKodeToad commented on 1.20/src/main/java/io/github/axolotlclient/mixin/PlayerEntityMixin.java:
```diff
@@ -84,4 +86,12 @@ public PlayerEntityMixin(EntityType<?> entityType, World world) {
 	}
+
+	@Override
+	public int getArmor() {
```
assuming this is for the hud, i'd recommend modifying the hud

moehreag replied:
yes

TheKodeToad commented on 1.20/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsGame.java:
```diff
@@ -0,0 +1,441 @@
+
+	private BedwarsPlayer me = null;
+
+	private final Map<String, BedwarsPlayer> players = new HashMap<>(16);
+	private final Map<UUID, BedwarsPlayer> playersById = new HashMap<>(16);
```
```suggestion
	private final Map<String, BedwarsPlayer> players = new HashMap<>();
	private final Map<UUID, BedwarsPlayer> playersById = new HashMap<>();
```
16 is already the default

moehreag replied:
true.

TheKodeToad reviewed (commented):
this shouldn't be considered a complete review, however i somewhat care about my sanity :D

moehreag committed:
address changes requested by TheKodeToad

moehreag committed:
fix 1.19.4 mixin

moehreag committed:
bump version and add changelog

moehreag committed:
remove code duplication

moehreag committed:
fix the ArmorHUD on 1.20

moehreag committed:
update config version to fix a bug

- involving the chroma switch
- https://github.com/AxolotlClient/AxolotlClient-config/commit/0bca1721525f0e1b1260bc73883eb6893900e871

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsMod.java:
```diff
@@ -0,0 +1,230 @@
+	protected final BooleanOption removeAnnoyingMessages = new BooleanOption(getTranslationKey("removeAnnoyingMessages"), true);
+
+
+	private final BooleanOption tabRenderLatencyIcon = new BooleanOption(getTranslationKey("tabRenderLatencyIcon"), false);
```
It may be worthwhile to add to the description that in hypixel games everyone's ping is sent as 1

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/mixin/PlayerListHudMixin.java:
```diff
@@ -150,4 +165,156 @@ public abstract class PlayerListHudMixin {
+		ci.cancel();
+		String render;
+		int color;
+		if (!bedwarsPlayer.isAlive()) {
```
Perhaps move this into BedwarsGame so it's easier to change in the future?

moehreag replied:
ok!

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/mixin/PlayerListHudMixin.java:
```diff
@@ -150,4 +165,156 @@ public abstract class PlayerListHudMixin {
+			value -= 9;
+		}
+		if (BedwarsMod.getInstance().isEnabled() && BedwarsMod.getInstance().isWaiting()) {
+			value += 20;
```
This is a magic number for making sure the stars display well enough. This could maybe be changed into an actual accurate width value.

moehreag replied:
that's not really possible without jank.. (since we don't have access to any string here). It's also an issue with the NumericalPing option though.

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/mixin/PlayerListHudMixin.java:
```diff
@@ -150,4 +165,156 @@ public abstract class PlayerListHudMixin {
+				return;
+			}
+
+			render = String.valueOf(HypixelAbstractionLayer.getPlayerLevel(playerListEntry2
```
This should probably be configurable and depending on option display the star icon.

moehreag replied:
What do you mean exactly?

DarkKronicle replied:
Ability for the user to change what stat it displays. Also, if it displays stars (bedwars level) I think a star icon would make it easier to tell what it is. Maybe even have basic prestige colors, but that's not necessary.

moehreag replied:
Should it just use the already existing option from LevelHead or should it have its own option to set the mode? Or do you mean it should just use the data from the Bedwars LevelHead?

DarkKronicle replied:
I think own option just for Bedwars. Just so that you can better see stars or like winstreaks if you want

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/autogg/AutoGG.java:
```diff
@@ -158,28 +161,28 @@ private List<String> addToList(String... strings) {
 			for (String s : options) {
-				if (messageReceived.getString().contains(s)) {
+				if (messageReceived.contains(s)) {
 					Util.sendChatMessage(messageToSend);
```
If on hypixel this needs to prepend `/achat ` so it will always send in public chat and not in party/guild chat.

moehreag replied:
since the string sent in the chat is only defined in an option this is user-configurable and can be configured to suit whatever purpose the user intends.

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/BedwarsMod.java:
```diff
@@ -0,0 +1,230 @@
+	private final static Pattern[] GAME_START = {
+		Pattern.compile("^\\s*?Protect your bed and destroy the enemy beds\\.\\s*?$"),
+		Pattern.compile("^\\s*?Bed Wars Lucky Blocks\\s*?$"),
+		Pattern.compile("^\\s*?Bed Wars Swappage\\s*?$")
```
Have to figure out which dreams mode the mod should initiate in. Swappage should almost certainly be removed since team swaps will completely break the mod. This was originally added just to provide basic testing.

moehreag replied:
I'd assume that all modes work except ones where teams are mixed up (which means just Swappage currently)

DarkKronicle replied:
I think only swappage and lucky blocks won't work. Lucky blocks has a placeable bed.

moehreag replied:
true, I forgot about that.

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/stats/LobbyStatsHud.java:
```diff
@@ -0,0 +1,53 @@
+ * @author DarkKronicle
+ */
+
+public class LobbyStatsHud extends BoxHudEntry {
```
Remove this file

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/upgrades/BedwarsTeamUpgrades.java:
```diff
@@ -0,0 +1,100 @@
+ * @author DarkKronicle
+ */
+
+public class BedwarsTeamUpgrades {
```
Would it be possible to check for a texture within your resource pack and then default to vanilla ones if none are found? This would allow customizing via a resource pack.

moehreag replied:
already possible by default, this implementation is enough for that

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/upgrades/BedwarsTeamUpgrades.java:
```diff
@@ -0,0 +1,100 @@
+ * @author DarkKronicle
+ */
+
+public class BedwarsTeamUpgrades {
```
I need to test how these look in game.

DarkKronicle commented on 1.16.5/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/upgrades/TrapUpgrade.java:
```diff
@@ -0,0 +1,130 @@
+public class TrapUpgrade extends TeamUpgrade {
+
+	private final static Pattern[] REGEX = {
+		Pattern.compile("^\\b[A-Za-z0-9_§]{3,16}\\b purchased (.+) Trap\\s*$"),
```
Remove redundant ones and I also don't think `It's a trap!` gets triggered. But I can never remember how the regex should go for that one.

moehreag replied:
I'm not entirely sure either..

moehreag committed:
address suggestions from DarkKronicle

moehreag committed:
Merge branch 'multiversion' into feat/sol/darkkronicle-bedwars-display

moehreag committed:
remove Lucky Block Bedwars from start triggers

moehreag committed:
fix last killer display

moehreag committed:
fix a mixin on 1.16 versions

moehreag committed:
add random stat generation

- lower the request limit due to new API rate limits

moehreag committed:
clamp randomly generated stats

- update Trap RegEx

moehreag committed:
Merge branch 'multiversion' into feat/sol/darkkronicle-bedwars-display
